src/Datadog.Trace/Sampling/RuleBasedSampler.cs

```diff
@@ -0,0 +1,134 @@
+using System;
+using System.Collections.Generic;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.Sampling
+{
+    internal class RuleBasedSampler : ISampler
+    {
+        private const ulong KnuthFactor = 1_111_111_111_111_111_111;
+
+        private static readonly Vendors.Serilog.ILogger Log = DatadogLogging.For<RuleBasedSampler>();
+
+        private readonly IRateLimiter _limiter;
+        private readonly List<ISamplingRule> _rules = new List<ISamplingRule>();
+
+        private readonly object _sampleRateGate = new object();
+
+        private Dictionary<string, float> _sampleRates = new Dictionary<string, float>();
+
+        public RuleBasedSampler(IRateLimiter limiter)
+        {
+            _limiter = limiter ?? new RateLimiter(null);
+        }
+
+        public void SetDefaultSampleRates(IEnumerable<KeyValuePair<string, float>> sampleRates)
+        {
+            // to avoid locking if writers and readers can access the dictionary at the same time,
+            // build the new dictionary first, then replace the old one
+            var rates = new Dictionary<string, float>(StringComparer.OrdinalIgnoreCase);
+
+            if (sampleRates != null)
+            {
+                foreach (var pair in sampleRates)
+                {
+                    rates.Add(pair.Key, pair.Value);
+                }
+            }
+
+            lock (_sampleRateGate)
```

Comment:
Good addition 👍

Reply:
Do we need any locking? The field assignment here is atomic, the dictionary itself is never modified, and the only read is in line 76 (`_sampleRates.TryGetValue(...)`).

Reply:
It can theoretically be re-assigned, seemed safer to me. Do you see harm in it?

Reply:
If it can't be reset, I'd like to move the assignment to the constructor, or put some safety code into the Set method.

Reply:
> It can theoretically be re-assigned, seemed safer to me.

Assigning a reference is an atomic operation, so this is safe. Code reading the reference will get either the old value or the new one (depending on timing), but never an invalid "in-between" value. Adding the locks won't change this.

It would be a different issue if we were adding key/value pairs to the dictionary while other threads were reading them. In that case, I would recommend ` ConcurrentDictionary` over locking, since `ConcurrentDictionary` implements reads without the use of locks. But that's not what happens here. We add all the values to a local dictionary variable `rates` that no other thread has access to, and only assign it to the field once we're done modifying it.

> Do you see harm in it?

Yes, locks aren't free. Adding locks here adds overhead for zero benefit.

> I'd like to move the assignment to the constructor

I like this idea! Sadly, the legacy sampling values that come from the Agent change over the life of the `Tracer`. We potentially get new sampling values every time we send Spans to the Agent.